File: learn/dynamic-programming/digit-dp/digit_dp.go


import "strconv"

type DigitDP struct {
	s    string
	memo [][]int64
}

func NewDigitDP() *DigitDP {
	return &DigitDP{}
}

func (dp *DigitDP) solve(id int, flag bool) int64 {
	if id == len(dp.s) {
		return 1
	}

	flagIdx := 0
	if flag {
		flagIdx = 1
	}

	if dp.memo[id][flagIdx] != -1 {
		return dp.memo[id][flagIdx]
	}

	var ans int64 = 0
	limit := 9
	if flag {
		limit = int(dp.s[id] - '0')
	}

	for i := 0; i <= limit; i++ {
		newFlag := flag && (i == limit)
		ans += dp.solve(id+1, newFlag)
	}

	dp.memo[id][flagIdx] = ans
	return ans
}

func (dp *DigitDP) CountValid(target int64) int64 {
	if target < 0 {
		return 0
	}
	dp.s = strconv.FormatInt(target, 10)
	dp.memo = make([][]int64, len(dp.s))
	for i := range dp.memo {
		dp.memo[i] = []int64{-1, -1}
	}
	return dp.solve(0, true)
}

func (dp *DigitDP) CountInRange(L, R int64) int64 {
	return dp.CountValid(R) - dp.CountValid(L-1)
}

File: learn/dynamic-programming/digit-dp/digit_dp.rs
pub struct DigitDP {
    s: String,
    memo: Vec<Vec<i64>>,
}

impl DigitDP {
    pub fn new() -> Self {
        DigitDP {
            s: String::new(),
            memo: Vec::new(),
        }
    }

    fn solve(&mut self, id: usize, flag: bool) -> i64 {
        if id == self.s.len() {
            return 1;
        }

        let flag_idx = if flag { 1 } else { 0 };

        if self.memo[id][flag_idx] != -1 {
            return self.memo[id][flag_idx];
        }

        let mut ans: i64 = 0;
        let limit = if flag {
            self.s.as_bytes()[id] - b'0'
        } else {
            9
        };

        for i in 0..=limit {
            let new_flag = flag && (i == limit);
            ans += self.solve(id + 1, new_flag);
        }

        self.memo[id][flag_idx] = ans;
        ans
    }

    pub fn count_valid(&mut self, target: i64) -> i64 {
        if target < 0 {
            return 0;
        }
        self.s = target.to_string();
        self.memo = vec![vec![-1; 2]; self.s.len()];
        self.solve(0, true)
    }

    pub fn count_in_range(&mut self, l: i64, r: i64) -> i64 {
        self.count_valid(r) - self.count_valid(l - 1)
    }
}
